pages/login.vue

```diff
@@ -1,3 +1,85 @@
+<script setup>
+    import pkceChallenge from 'pkce-challenge';
+
+    const client_id = "20aa48c2719e42c0be5f3b834942f06d";
+    const scopes = [];  // https://developer.spotify.com/documentation/general/guides/authorization/scopes/
+    const redirect_uri = new URL(window.location);  // use the current location
+    redirect_uri.search = "";  // clear query parameters
+    const auth_token = localStorage.getItem('auth-token');
+
+    async function logout(){
+        localStorage.removeItem('auth-token');
+        navigateTo('/');
+    }
+
+    async function login(){
+        const pkce = pkceChallenge();
+
+        // save for pkce post-request
+        localStorage.setItem('code-verifier', pkce.code_verifier);
+
+        const url = new URL('/authorize', 'https://accounts.spotify.com/');
+        url.searchParams.append('client_id', client_id);
+        url.searchParams.append('response_type', 'code');
+        url.searchParams.append('redirect_uri', redirect_uri);
+        url.searchParams.append('scopes', scopes.join(" "));
+        url.searchParams.append('code_challenge_method', 'S256');
+        url.searchParams.append('code_challenge', pkce.code_challenge);
+        navigateTo(url.toString(), {external: true});  // page redirect
+    }
+
+    async function loadToken(){
+        // generated in login()
+        const code_verifier = localStorage.getItem('code-verifier');
+
+        const url = new URL('/api/token', 'https://accounts.spotify.com/');
+
+        const form = {
+            grant_type: "authorization_code",
+            code: pkce,
+            redirect_uri: redirect_uri,
+            client_id: client_id,
+            code_verifier: code_verifier
+        };
+        
+        const response = await fetch(url, {
+            method: "POST",
+            body: new URLSearchParams(form),
+            headers: {
+                "Content-Type": "application/x-www-form-urlencoded"
+            }
+        });
+
+        const data = await response.json();
```

Comment:
ich glaube wir können line 45 und 53 auf eine Zeile kombinieren. Dann haben wir einen await weniger

Reply:
nein. du meckerst immer wegen ordnung und so. und auf diese art ist die request und das verarbeiten der request getrennt

Reply:
mich verwirrt nur das zweite await. Meiner Meinung nach brauch man es an der zweiten stelle nicht, aber ich kenn ich in dem Bereich nicht aus. Deswegen hätte ich die lines kombiniert um sicher zu sein. Ansonsten find ich das gar nicht so schlecht. (vielleicht ist der name data etwas generisch)

Reply:
extra data wegen generisch gewählt, da er statt dem access_token auch error und error_description? sein kann